File: leetCode/238.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// 左右乘机列表
var productExceptSelf = function(nums) {
    let ans = new Array(nums.length).fill(1)
    let left = new Array(nums.length).fill(1)
    let right = new Array(nums.length).fill(1)
    for(let i = 1; i < nums.length; i++) {
        left[i] =  nums[i - 1] * left[i - 1]
    }
    for(let i = nums.length - 2; i >= 0; i--) {
        right[i] =  nums[i + 1] * left[i + 1]
    }

    for(let i = 0 ; i < nums.length; i++ ) {
        ans[i] = left[i] * right[i]
    }
    return ans
};
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// 采用除法 + 记录每一次合层的数据
var productExceptSelf1 = function(nums) {
    let answers = [1]
    for(let i = 1; i < nums.length; i++) answers[0] *= nums[i]
    for(let i = 1; i < nums.length; i++) {
        if (nums[i] === 0) {
            answers[i] = 1
            for (let j = 0; j < nums.length; j++) {
                if (j === i) continue;
                answers[i] *= nums[j]
            }
        } else {
            answers[i] = answers[i - 1] * nums[i - 1] / nums[i]
        }
    }

    return answers
};

File: leetCode/238.go
func productExceptSelf(nums []int) []int {
    size := len(nums)
    ans := make([]int,size)
    left := make([]int,size)
    right := make([]int,size)
    left[0] = 1
    right[size- 1] = 1
    for i := 1; i < len(nums); i++ {
        left[i] = left[i - 1] * nums[i - 1]
    }
        fmt.Println(left)
    for i := len(nums) - 2; i >= 0; i-- {
        right[i] = right[i + 1] * nums[i + 1]
    }

    for i,_ := range ans {
        ans[i] = left[i] * right[i]
    }

    return ans
}